Precompiled script plugins - `GradleProperties has not been loaded yet.`

Original issue: https://issuetracker.google.com/162515856
Gradle version: 6.6.0-rc-4

Using the [test.zip](https://github.com/gradle/gradle/files/5021871/test.zip) project, `:buildSrc:generatePrecompiledScriptPluginAccessors` task fails with:

```
Caused by: java.lang.IllegalStateException: GradleProperties has not been loaded yet.
        at org.gradle.initialization.DefaultGradlePropertiesController$NotLoaded.gradleProperties(DefaultGradlePropertiesController.java:74)
        at org.gradle.initialization.DefaultGradlePropertiesController$SharedGradleProperties.gradleProperties(DefaultGradlePropertiesController.java:59)
        at org.gradle.initialization.DefaultGradlePropertiesController$SharedGradleProperties.find(DefaultGradlePropertiesController.java:50)
        at org.gradle.api.internal.provider.sources.GradlePropertyValueSource.obtain(GradlePropertyValueSource.java:36)
        at org.gradle.api.internal.provider.sources.GradlePropertyValueSource.obtain(GradlePropertyValueSource.java:24)
        at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$LazilyObtainedValue.lambda$obtainValueForThe1stTime$0(DefaultValueSourceProviderFactory.java:344)
        at org.gradle.internal.Try.ofFailable(Try.java:39)
        at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$LazilyObtainedValue.obtainValueForThe1stTime(DefaultValueSourceProviderFactory.java:344)
        at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$LazilyObtainedValue.obtain(DefaultValueSourceProviderFactory.java:333)
        at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$ValueSourceProvider.calculateOwnValue(DefaultValueSourceProviderFactory.java:299)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.getOrNull(AbstractMinimalProvider.java:93)
        at com.android.build.gradle.options.ProjectOptions.readTestRunnerArgs(ProjectOptions.java:153)
        at com.android.build.gradle.options.ProjectOptions.<init>(ProjectOptions.java:62)
        at com.android.build.gradle.options.ProjectOptions.<init>(ProjectOptions.java:76)
        at com.android.build.gradle.internal.plugins.BasePlugin.createProjectServices(BasePlugin.java:846)
        at com.android.build.gradle.internal.plugins.BasePlugin.basePluginApply(BasePlugin.java:255)
        at com.android.build.gradle.internal.plugins.BasePlugin.lambda$apply$0(BasePlugin.java:245)
        at com.android.build.gradle.internal.crash.CrashReporting.runAction(crash_reporting.kt:27)
        at com.android.build.gradle.internal.plugins.BasePlugin.apply(BasePlugin.java:243)
        at com.android.build.gradle.internal.plugins.BasePlugin.apply(BasePlugin.java:138)
        at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:43)
        at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)
        at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:177)
        at org.gradle.api.internal.plugins.DefaultPluginManager.access$100(DefaultPluginManager.java:51)
        at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:272)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:395)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:387)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:84)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.plugins.DefaultPluginManager.lambda$doApply$0(DefaultPluginManager.java:157)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:43)
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:156)
```

Is `GradlePropertiesController` supposed to load properties earlier?

--- a/subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java
+++ b/subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java
@@ -42,6 +42,10 @@ public class DefaultGradlePropertiesController implements GradlePropertiesContro
         state = state.loadGradlePropertiesFrom(settingsDir);
     }
 
+    public void overrideWith(GradleProperties gradleProperties) {
+        state = state.overrideWith(gradleProperties);
+    }
+
     private class SharedGradleProperties implements GradleProperties {
 
         @Nullable
@@ -65,6 +69,8 @@ public class DefaultGradlePropertiesController implements GradlePropertiesContro
         GradleProperties gradleProperties();
 
         State loadGradlePropertiesFrom(File settingsDir);
+
+        State overrideWith(GradleProperties gradleProperties);
     }
 
     private class NotLoaded implements State {
@@ -80,6 +86,11 @@ public class DefaultGradlePropertiesController implements GradlePropertiesContro
                 propertiesLoader.loadGradleProperties(settingsDir),
                 settingsDir
             );
+        }
+
+        @Override
+        public State overrideWith(GradleProperties gradleProperties) {
+            return new Overridden(gradleProperties);
         }
     }
 
@@ -109,6 +120,35 @@ public class DefaultGradlePropertiesController implements GradlePropertiesContro
                 );
             }
             return this;
+        }
+
+        @Override
+        public State overrideWith(GradleProperties gradleProperties) {
+            throw new IllegalStateException("GradleProperties has already been loaded and cannot be overridden.");
+        }
+    }
+
+    private static class Overridden implements State {
+
+        private final GradleProperties gradleProperties;
+
+        public Overridden(GradleProperties gradleProperties) {
+            this.gradleProperties = gradleProperties;
+        }
+
+        @Override
+        public GradleProperties gradleProperties() {
+            return gradleProperties;
+        }
+
+        @Override
+        public State loadGradlePropertiesFrom(File settingsDir) {
+            throw new IllegalStateException();
+        }
+
+        @Override
+        public State overrideWith(GradleProperties gradleProperties) {
+            return new Overridden(gradleProperties);
         }
     }
 }

--- a/subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPluginAccessorsTest.kt
+++ b/subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPluginAccessorsTest.kt
@@ -21,6 +21,8 @@ import com.nhaarman.mockito_kotlin.doReturn
 import com.nhaarman.mockito_kotlin.inOrder
 import com.nhaarman.mockito_kotlin.mock
 
+import org.codehaus.groovy.runtime.StringGroovyMethods
+
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.plugins.PluginManager
@@ -59,6 +61,71 @@ import java.io.File
 
 @LeaksFileHandles("Kotlin Compiler Daemon working directory")
 class PrecompiledScriptPluginAccessorsTest : AbstractPrecompiledScriptPluginTest() {
+
+    @ToBeFixedForInstantExecution
+    @Test
+    fun `can use type-safe accessors for plugin relying on gradleProperty provider`() {
+
+        withFolders {
+            "buildSrc" {
+                "gradlePropertyPlugin" {
+                    withKotlinDslPlugin()
+                    withFile("src/main/kotlin/gradlePropertyPlugin.gradle.kts", """
+                        val property = providers.gradleProperty("theGradleProperty").forUseAtConfigurationTime()
+                        if (property.isPresent) {
+                            println("property is present in plugin!")
+                        }
+                        extensions.add<Provider<String>>(
+                            "theGradleProperty",
+                            property
+                        )
+                    """)
+                }
+                "gradlePropertyPluginConsumer" {
+                    withKotlinDslPlugin().appendText("""
+                        dependencies {
+                            implementation(project(":gradlePropertyPlugin"))
+                        }
+                    """)
+                    withFile("src/main/kotlin/gradlePropertyPluginConsumer.gradle.kts", """
+                        plugins { id("gradlePropertyPlugin") }
+
+                        if (theGradleProperty.isPresent) {
+                            println("property is present in consumer!")
+                        }
+                    """)
+                }
+
+                withDefaultSettingsIn(relativePath).appendText("""
+                    include("gradlePropertyPlugin")
+                    include("gradlePropertyPluginConsumer")
+                """)
+                withBuildScriptIn(relativePath, """
+                    plugins { `java-library` }
+                    dependencies {
+                        subprojects.forEach {
+                            runtimeOnly(project(it.path))
+                        }
+                    }
+                """)
+            }
+        }
+
+        withBuildScript("""
+            plugins { gradlePropertyPluginConsumer }
+        """)
+
+        build("help", "-PtheGradleProperty=42").apply {
+            assertThat(
+                output.count("property is present in plugin!"),
+                equalTo(1) // not printed when building buildSrc
+            )
+            assertThat(
+                output.count("property is present in consumer!"),
+                equalTo(1)
+            )
+        }
+    }
 
     @ToBeFixedForInstantExecution
     @Test
@@ -653,4 +720,8 @@ class PrecompiledScriptPluginAccessorsTest : AbstractPrecompiledScriptPluginTest
         }
         return Pair(project, pluginManager)
     }
+
+    private
+    fun CharSequence.count(text: CharSequence): Int =
+        StringGroovyMethods.count(this, text)
 }

--- a/subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/tasks/GeneratePrecompiledScriptPluginAccessors.kt
+++ b/subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/tasks/GeneratePrecompiledScriptPluginAccessors.kt
@@ -20,10 +20,12 @@ import org.gradle.api.Project
 import org.gradle.api.file.ConfigurableFileCollection
 import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.file.ProjectLayout
+import org.gradle.api.internal.GradleInternal
 import org.gradle.api.internal.artifacts.dependencies.DefaultSelfResolvingDependency
 import org.gradle.api.internal.file.FileCollectionFactory
 import org.gradle.api.internal.initialization.ScriptHandlerInternal
 import org.gradle.api.internal.project.ProjectInternal
+import org.gradle.api.internal.properties.GradleProperties
 import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.Classpath
 import org.gradle.api.tasks.InputDirectory
@@ -35,6 +37,8 @@ import org.gradle.api.tasks.PathSensitivity
 import org.gradle.api.tasks.TaskAction
 import org.gradle.groovy.scripts.TextResourceScriptSource
 import org.gradle.initialization.ClassLoaderScopeRegistry
+import org.gradle.initialization.DefaultGradlePropertiesController
+import org.gradle.initialization.GradlePropertiesController
 import org.gradle.internal.classpath.DefaultClassPath
 import org.gradle.internal.concurrent.CompositeStoppable.stoppable
 import org.gradle.internal.exceptions.LocationAwareException
@@ -254,7 +258,10 @@ abstract class GeneratePrecompiledScriptPluginAccessors @Inject internal constru
     private
     fun scriptSourceFor(plugin: PrecompiledScriptPlugin) =
         TextResourceScriptSource(
-            textFileResourceLoader.loadFile("Precompiled script plugin", plugin.scriptFile)
+            textFileResourceLoader.loadFile(
+                "Precompiled script plugin",
+                plugin.scriptFile
+            )
         )
 
     private
@@ -392,12 +399,31 @@ class SyntheticProjectSchemaBuilder(
             .withGradleUserHomeDir(gradleUserHomeDir)
             .withProjectDir(projectDir)
             .build()
+            .withEmptyGradleProperties()
 
         addScriptClassPathDependencyTo(project, rootProjectClassPath)
 
         applyPluginsTo(project, PluginRequests.EMPTY)
 
         return project
+    }
+
+    private
+    fun Project.withEmptyGradleProperties(): Project {
+        gradle.run {
+            require(this is GradleInternal)
+            services[GradlePropertiesController::class.java].run {
+                require(this is DefaultGradlePropertiesController)
+                overrideWith(EmptyGradleProperties)
+            }
+        }
+        return this
+    }
+
+    private
+    object EmptyGradleProperties : GradleProperties {
+        override fun find(propertyName: String?) = null
+        override fun mergeProperties(properties: Map<String, String>) = properties.toMap()
     }
 
     private

